### PR TITLE
Integrate sqlparser

### LINKED ISSUE #321: Add a query parser
This issue tracks the effort of adding a parser for SQLite queries, which will enable us to do static analysis on the queries, get dependencies of DatabaseViews, be aware of affected databases when allowing arbitrary queries and much more. 

I have started implementing this for a prototype and want to build on the work done in #320 .

The current strategy is to use the antlr SQLite grammar updated [here](https://github.com/bkiers/sqlite-parser/pull/22)(MIT licensed) and use the dart target of antlr [as soon as it is ready](https://github.com/antlr/antlr4/pull/2728):
- Generate the parser files once and add them into the git repository because they don't have to be generated frequently and we will need to have the antlr tool installed for it
  - Document the generation process/commands
  - Let dartanalyzer ignore the generated files
- Add the dart antlr runtime as a dependency on `floor_generator`

The plan is then to give `QueryAdapter.query(List)Stream` a set of entities to watch out for (instead of the single entity it waits on right now). I'm not sure how to wire it up to there yet but it should not be complicated.

I'll open a PR once I have a prototype ready. At the moment this includes a git submodule containing the antlr PR referenced above and I don't want to push that.

### PR DESCRIPTION
**Pull request by mqus**
I have now finished an early working version of the integrated sqlparser. There are still some `//todo`s open, most of them related to the typeConverters, because I assume that this PR will not be merged before #318 (though mostly because the query parser is not quite finished).

### Features :
- An even more intelligent generator overall and the foundation for more features
- Query validation for `@DatabaseView` and `@Query`. (#58)
- Type derivation and matching for `@DatabaseView` and `@Query`.
- Dependency and "What entities are affected" derivation, including foreign keys of entities
- Better parameter handling 
  - automatic reordering
  - multiple insertions of the same parameter (#142)
  - better handling of list parameters (fixing sql-injections)
  - removed string-replace functions which could mess up content in string literals
- Outputting simple values in Queries (#200)
- Cheaper streams on DatabaseViews

### TODO:
**Implement some Features/Refactor**
- [x] Mapper for primitive types
- [x] Return type checker
- [x] Restructure  AnalyzeResult creation
- [x] Maybe separate type checker from view converter
- [x] Check view query for variables
- [x] Map bool type to int with typehint in converters
- [x] Make listInsertionPositions a (sorted) list with dedicated type
- [x] Formulate missing errors
- [ ] Check for and add missing errors

**Cleanup**
- [x] Check todo comments for missing functionality and expand this todo list
- [x] Add `@nonNull`/`@nullable` annotations to new functions and check usage
- [x] Add Docstrings to crucial functions

**Integration of other PRs**
- [ ] integrating typeconverters
- [ ] Maybe integrate #359, if it is merged.

**Fixing existing tests**
- [x] integration tests (pretty much already worked without any changes)
- [x] queryadapter unit tests
- [x] generator unit tests

**Adding new tests**
- [ ] integration tests
- [x] queryadapter unit tests
- [ ] generator unit tests
  - [x] Add regression test from #353

**Update README**
- [x] Document subtleties, such as executing a query only up to the first semicolon and not allowing more than 1000 parameter values (both limitations of the sqlite implementations of e.g. android)
- [x] Update Documentation regarding streams on Views

### Small Guide to the PR
I used the library sqlparser from https://github.com/simolus3/moor, like mentioned before.
I added a new subdirectory to `processor` called `query_analyzer`. There are multiple files in there:

- **engine.dart**/`AnalyzerEngine` is the functional wrapper for the sqlparser library, which is initialized once per database and holds the context for the analyzer. It stores dependencies, maps database names to Queryables and is able to analyze queries by wrapping sqlparsers engine. All entities and database views have to be registered here, with database views having to come after entities for a successful validation. 

- **dependency_graph.dart** is a small utility data structure for the engine to hold the dependency graph. It currently calculates the indirect dependencies without caching any results to keep it simple. There are no plans to include caching yet as generator performance is not a priority right now.

- **type_checker.dart** contains functions to compare the types of view fields/output fields of a query output to the derived youtput from the associated query and reports errors.

- **visitors.dart** Is a visitor which walks the sqlparser tree and stores all found variable references, such as `?`, `?44`, `:varname`, and so on. It is used by the QueryProcessor to first validate the parameters and then find and replace all named variables starting with a colon in the sql string with either a numbered variable (for normal parameters of the query) or a placeholder (for list parameters).

I also added **query.dart** and **query_processor.dart**, which are responsible for holding and processing the query, including the derivation of list parameter mappings, output types, dependencies and affected entities, as well as the validation of the query.

Finally, I also added a `QueryMethodReturnType` to the value_object folder to simplify the access to the return type description of query methods and restructured the `query_method_writer.dart`. 

If there are any questions or suggestions, I'm happy to listen, while I work on the tasks written above.

Closes #321

### CODE REVIEW

**Comment by vitusortner** [comment]
This looks very promising 👏 I'll make time for reviewing the changes over the next days. If there is anything in particular that I can help you with, just let me know!

**Comment by mqus** [comment]
In the meantime, let me give you a small taste of what we're getting with @simolus3's sqlparser. I currently fix the existing tests on the query_method writer and processor and many of the tests have some arbitrary queries in them, that don't reference any existing entities or wrong names, which weren't necessary to be right until now. This is the kind of error messages I get:
```
The query contained errors: Could not find Persons. Available are: sqlite_master, sqlite_sequence, Person, Name
package:_resolve_source/_resolve_source.dart:8:28
  ╷
8 │       Future<List<Person>> findPersonsWithNamesLike(String name);
  │                            ^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
```
I think they're awesome. I'll look into also printing the line before that which actually contains the query.

**Comment by mqus** [comment]
> If there is anything in particular that I can help you with, just let me know!

A list of testcases you would like to see covered, I have my own list but more is better :)

**Comment by vitusortner** [comment]
> In the meantime, let me give you a small taste of what we're getting with @simolus3's sqlparser. I currently fix the existing tests on the query_method writer and processor and many of the tests have some arbitrary queries in them, that don't reference any existing entities or wrong names, which weren't necessary to be right until now. This is the kind of error messages I get:
> 
> ```
> The query contained errors: Could not find Persons. Available are: sqlite_master, sqlite_sequence, Person, Name
> package:_resolve_source/_resolve_source.dart:8:28
>   ╷
> 8 │       Future<List<Person>> findPersonsWithNamesLike(String name);
>   │                            ^^^^^^^^^^^^^^^^^^^^^^^^
>   ╵
> ```
> 
> I think they're awesome. I'll look into also printing the line before that which actually contains the query.

That's amazing! The capabilities your changes provide to the project are more than great. Good job 👏

**Comment by mqus** [comment]
Short status update: feature-wise this is ready to use. I'm currently adding tests and making some minor refactors and bug fixes. The main Items left to do after the tests are mainly cleaning up work, such as docstrings for functions, (non)-nullable annotations.

Of course, this will have some major overlaps with the typeconverters (and some fairly minor ones(i hope) with `@embedded`) so there will be some work needed to combine this and add some more tests for the cross sections.

**Comment by mqus** [comment]
There seem to be quite a few people who want to use this in the stable release and I would now rather get this merged before the type converters, because they seem to need a bit longer. @vitusortner could you review the implementation part so far? I will continue adding tests but the rest is mostly complete. Alternatively we could also merge the `@embedded` PR first, if this was reviewed.

**Comment by vitusortner** [comment]
I'll look into it ASAP!

File: floor/test/adapter/query_adapter_test.dart [hunk 1]
```diff
@@ -282,7 +308,36 @@ void main() {
...
+          ]));
+
+      //after emitting the first result, no other updates were made
+      expect(actual.skip(1), emitsDone);
+
+      streamController.add(otherEntity);
+      streamController.add(otherEntity);
+
+      await Future<void>.delayed(const Duration(milliseconds: 100));
+      await streamController.close();
```
Comment (vitusortner) [inline floor/test/adapter/query_adapter_test.dart:340]: 💡 We can remove this line as we close `StreamController`s in the `tearDown` block after each test.

```suggestion
```

File: floor/test/integration/boolean_conversions/bool_test.dart [hunk 2]
```diff
@@ -64,6 +64,23 @@ void main() {
...
+
+    test('return nullables', () async {
+      await boolDao
+          .insertBoolC(BooleanClass(false, nullable: null, nonnullable: true));
+      await boolDao
+          .insertBoolC(BooleanClass(true, nullable: false, nonnullable: true));
+
+      expect(await boolDao.getNullables(), equals([null, false]));
+    });
+    test('return nullables', () async {
```
Comment (vitusortner) [inline floor/test/integration/boolean_conversions/bool_test.dart:76]: 💡  The test's name doesn't reflect the actual test.

```suggestion

    test('return nonNullables', () async {
```

File: floor/test/integration/sqlparser_test/sqlparser_test.dart [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    await Future<void>.delayed(const Duration(milliseconds: 100));
+    await caseDao.insertPerson(person1);
+
+    await Future<void>.delayed(const Duration(milliseconds: 100));
+    await caseDao.updateName();
+
+    await Future<void>.delayed(const Duration(milliseconds: 100));
+    await caseDao.insertPerson(person2);
+
+    await Future<void>.delayed(const Duration(milliseconds: 100));
```
Comment (vitusortner) [inline floor/test/integration/sqlparser_test/sqlparser_test.dart:54]: ❔ I guess without these delays, the tests don't run?

File: floor/test/integration/sqlparser_test/sqlparser_test.dart [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    await caseDao.updateName();
+  });
+
+  group('Function type derivation tests', () {
+    test('try IN with null list', () async {
+      final list = ['a', 'b', 'c', 'd', 'e', 'f', null];
+
+      final aInList = await deepDao.isXinList('a', list);
+      expect(aInList, equals(true));
```
Comment (vitusortner) [inline floor/test/integration/sqlparser_test/sqlparser_test.dart:63]: ❔ The SQLite database returns "a" in this case? Which use cases do you see for mapping a result (of not `INTEGER`) to true when the return type is `bool` ?

File: floor/test/integration/stream_query_test.dart [hunk 5]
```diff
@@ -137,16 +137,20 @@ void main() {
           actual,
           emitsInOrder(<List<Dog>>[
             [], // initial state,
+            [], // after inserting person1,
+            [], // after inserting person2,
             [dog1], // after inserting dog1
             [dog1], // after inserting dog2
-            //[], // after removing person1. Does not work because
-            // ForeignKey-relations are not considered yet (#321)
+            [], // after removing person1.
```
Comment (vitusortner) [inline floor/test/integration/stream_query_test.dart:144]: ❤️ Good job! This is great!

File: floor_generator/lib/misc/string_utils.dart [hunk 6]
```diff
@@ -1,6 +1,27 @@
...
+import 'package:strings/strings.dart';
+
 extension StringUtils on String {
   /// Makes the first letter of the supplied string [value] lowercase.
+  @nonNull
   String decapitalize() {
     return '${this[0].toLowerCase()}${substring(1)}';
   }
+
+  /// Makes the first letter of the supplied string [value] lowercase.
```
Comment (vitusortner) [inline floor_generator/lib/misc/string_utils.dart:11]:
```suggestion
  /// Makes the first letter of the supplied string [value] uppercase.
```

File: floor_generator/lib/misc/type_utils.dart [hunk 7]
```diff
@@ -4,7 +4,9 @@ import 'package:analyzer/dart/constant/value.dart';
 import 'package:analyzer/dart/element/element.dart';
 import 'package:analyzer/dart/element/type.dart';
 import 'package:floor_generator/misc/annotations.dart';
+import 'package:floor_generator/misc/constants.dart';
 import 'package:source_gen/source_gen.dart';
+import 'package:sqlparser/sqlparser.dart' show BasicType;
```
Comment (vitusortner) [inline floor_generator/lib/misc/type_utils.dart:9]: ❤️ You really know got to know Dart in the meanwhile!

File: floor_generator/pubspec.yaml [hunk 8]
```diff
@@ -17,8 +17,15 @@ dependencies:
...
   collection: ^1.14.11
+  strings: ^0.1.2
   floor_annotation:
     path: ../floor_annotation/
+  sqlparser: # should be ^0.11.0
+    git:
+      url: https://github.com/simolus3/moor.git
+      ref: c2ce934
+      path: sqlparser
+
```
Comment (simolus3) [inline floor_generator/pubspec.yaml:28]: FYI: I just released version `0.10.1` which contains the commit referenced here. There weren't any breaking changes, which is why it's not `0.11.0`

Comment (mqus) [reply floor_generator/pubspec.yaml:28]: ah well, this was just to remind me to use the next version, not 0.11.0 explicitly. Thank you for the notification!

File: floor_generator/lib/processor/query_analyzer/engine.dart [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    registry[entity.name] = entity;
+
+    //register dependencies
+    final directDependencies = entity.foreignKeys
+        .where((e) => e.canChangeChild)
+        .map((e) => e.parentName);
+    dependencies.add(entity.name, directDependencies);
+  }
+
+  void registerView(floor.View floorView, View convertedView) {
```
[the remark this answers is not in the document]
Comment (mqus) [reply floor_generator/lib/processor/query_analyzer/engine.dart:46]: I thought about this and while entities can be converted without any fuzz (they don't have to be analyzed and can't throw any errors), Views have their query checked, have to be typechecked, can throw errors (which need the source reference). I had both solutions in the code at some time:
1. Add all entities and views to the engine at the `DatabaseProcessor`-level, after processing them.
   This has the "downside" that the conversion of the views is just treated like a conversion for the engine, while it really is the second part of the processing, throwing errors and determining dependencies.
2. Add views and entities to the engine, as the last processing step inside their `process()` methods.
   This solved the view issues in 1., but introduced the engine to the entity processor, which didn't really need it. It also made the tests for entities a lot more complex for gaining nothing in return.

So I chose to really handle them differently. This has it's own issues, like also having to treat them differently in tests, but I think that the tradeoffs are ok. Of course, this is just my opinion and I'm also not that sure on what the right way is. If you have a clear preference I'll gladly change this.

File: README.md [hunk 10]
```diff
@@ -421,9 +416,8 @@ abstract class PersonDao {
 
 ### Queries
 Method signatures turn into query methods by adding the `@Query()` annotation with the query in parenthesis to them.
-Be patient about the correctness of your SQL statements.
-They are only partly validated while generating the code.
-These queries have to return either a `Future` or a `Stream` of an entity or `void`.
+Your SQL queries will be validated completely while generating the code.
+These queries have to return either a `Future` or a `Stream` of an entity, a view, a primitive value or `void`.
```
[the remark this answers is not in the document]
Comment (mqus) [reply README.md:420]: This will need another PR but it should be as simple as adding another Queryable subtype and using the same processing as for entities/views. The type checking and the mapping already exists.

File: floor/test/integration/sqlparser_test/sqlparser_test.dart (hunk 3 above)
Comment (mqus) [reply floor/test/integration/sqlparser_test/sqlparser_test.dart:54]: Yeah, they will not complete successfully and will have issues like the following:
```
  Expected: should do the following in order:
            • emit an event that []
            • emit an event that []
            • emit an event that [Person:Person{id: 1, name: Baba}]
            • emit an event that [Person:Person{id: 1, name: Baba}]
            • emit an event that [Person:Person{id: 1, name: Baba}, Person:Person{id: 2, name: Me}]
            • emit an event that [Person:Person{id: 1, name: Baba}, Person:Person{id: 2, name: Me}]
    Actual: <Instance of '_BroadcastStream<List<Person>>'>
     Which: emitted • []
                    • [Person{id: 1, name: Baba}]
              which didn't emit an event that [] because it emitted an event that longer than expected at location [0]
```
Because the first two (or three) statements were completed together before the stream got to re-query the database.

File: floor/test/integration/stream_query_test.dart (hunk 5 above)
Comment (mqus) [reply floor/test/integration/stream_query_test.dart:144]: Here I found out that this is still suboptimal, since it will be also triggered if a person is inserted, even if that will never change `dog`. :laughing: 

This is what I meant in #373 .

File: floor/test/integration/sqlparser_test/sqlparser_test.dart (hunk 4 above)
Comment (mqus) [reply floor/test/integration/sqlparser_test/sqlparser_test.dart:63]: In my case it returns `1` and not `"a"`. This is also what I want to test: a simple function that does not refer to any table and returns a primitive boolean (or null) value.
```
sqlite> SELECT "a" IN ("a", "2", "s");
1
```
`isXinList` is just a complicated way to write `['a', 'b', 'c', 'd', 'e', 'f', null].contains('a')`.

File: floor/test/adapter/query_adapter_test.dart (hunk 1 above)
Comment (mqus) [reply floor/test/adapter/query_adapter_test.dart:340]: If I remove that, the test will fail because of a deadlock/timeout, since the `emitsInOrder` still waits for the Stream to complete, but the Stream only gets closed *after* the test succeeded.

File: floor/test/integration/boolean_conversions/bool_test.dart (hunk 2 above)
Comment (mqus) [reply floor/test/integration/boolean_conversions/bool_test.dart:76]: will fix!

File: floor_generator/lib/misc/string_utils.dart (hunk 6 above)
Comment (mqus) [reply floor_generator/lib/misc/string_utils.dart:11]: will fix!

File: floor_generator/lib/misc/type_utils.dart (hunk 7 above)
Comment (mqus) [reply floor_generator/lib/misc/type_utils.dart:9]: Well I added a `View` class to both sqlparser and floor and now I have to cope with the consequences :smile:

File: floor_generator/lib/processor/query_analyzer/engine.dart [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+@nonNull
+EngineOptions getDefaultEngineOptions() => EngineOptions(
+      useMoorExtensions: false,
+      useLegacyTypeInference: false,
+      enabledExtensions: const [],
+    );
+
+class AnalyzerEngine {
+  final Map<String, Queryable> registry = {};
```
[the remark this answers is not in the document]
Comment (mqus) [reply floor_generator/lib/processor/query_analyzer/engine.dart:26]: The main problem is that knownResultSets only stores sqlparser objects, not floor `Queryable`s. But this is only used in the QueryProcessor to find the matching entity to a dependency, which is then used to build the `streamEntities` Set of the Dao and the database. But both would be fine to only use the database name, which is unique. So I can definitely simplify this and a lot of other stuff and then remove this field completely :+1:

File: floor_generator/lib/processor/query_analyzer/engine.dart [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+  final SqlEngine inner = SqlEngine(getDefaultEngineOptions());
```
[the remark this answers is not in the document]
Comment (mqus) [reply floor_generator/lib/processor/query_analyzer/engine.dart:28]: will fix!

File: floor_generator/lib/processor/query_analyzer/engine.dart [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+  final DependencyGraph dependencies = DependencyGraph();
```
[the remark this answers is not in the document]
Comment (mqus) [reply floor_generator/lib/processor/query_analyzer/engine.dart:30]: will fix!

File: floor_generator/lib/processor/query_analyzer/engine.dart [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+  AnalyzerEngine();
```
[the remark this answers is not in the document]
Comment (mqus) [reply floor_generator/lib/processor/query_analyzer/engine.dart:32]: will fix!

**Comment by mqus** [comment]
`pub run test_cov` is failing, probably due to https://github.com/flutter/flutter/issues/57864

**Comment by mqus** [comment]
I think that there are enough conflicts by now that this has to be rewritten from scratch. 

I'm currently not up for it but if I have some advice for people who will try this in the future: take smaller steps. I rewrote some logic and also added a parser which already did 90% of the things it can do, all in one PR. This is too much to be reviewed reasonably, which is another reason why I think this PR has no hope of getting merged as-is, even when fixing all conflicts and updating it.

For those reasons, I will close this PR. Maybe I will have some time to go back and try again but currently I don't. I will gladly to give advice and reviews for future attempts of contributors, just ping me ;)
